### PR TITLE
Add more data into the execution trace

### PR DESCRIPTION
**Pull request by mask-pp**

### CODE REVIEW

File: core/blockchain.go [hunk 1]
```diff
@@ -1333,6 +1340,36 @@ func (bc *BlockChain) writeBlockWithState(block *types.Block, receipts []*types.
 	return status, nil
 }
 
+// Fill blockResult content
+func (bc *BlockChain) fillBlockResult(state *state.StateDB, block *types.Block, blockResult *types.BlockResult) {
```
Comment (0xmountaintop) [inline core/blockchain.go]: this function has a side-effect of writing BlockResult, so we should either:

+ separate it into 2 functions as `fillBlockResult` and `writeBlockResult`, or
+ rename it to `fillAndWriteBlockResult`

File: core/blockchain_reader.go [hunk 2]
```diff
@@ -158,6 +158,13 @@ func (bc *BlockChain) GetBlockByHash(hash common.Hash) *types.Block {
 	return bc.GetBlock(hash, *number)
 }
 
+func (bc *BlockChain) BlockResultByHash(blockHash common.Hash) *types.BlockResult {
```
Comment (0xmountaintop) [inline core/blockchain_reader.go]: `GetBlockResultByHash`

File: eth/api.go [hunk 3]
```diff
@@ -620,5 +620,8 @@ func NewPublicTraceAPI(eth *Ethereum) *PublicTraceAPI {
 
 // BlockResultByHash returns the blockResult by blockHash.
 func (api *PublicTraceAPI) BlockResultByHash(blockHash common.Hash) (*types.BlockResult, error) {
-	return rawdb.ReadBlockResult(api.e.chainDb, blockHash), nil
+	if blockResult := api.e.blockchain.BlockResultByHash(blockHash); blockResult != nil {
+		return blockResult, nil
+	}
+	return nil, fmt.Errorf("no block result found")
```
Comment (0xmountaintop) [inline eth/api.go]: `GetBlockResultByHash`

File: core/vm/logger.go [hunk 4]
```diff
@@ -200,6 +206,72 @@ func (l *StructLogger) CaptureState(pc uint64, op OpCode, gas, cost uint64, scop
...
+	if !l.cfg.DisableStack {
+		stck = make([]uint256.Int, len(stack.Data()))
+		for i, item := range stack.Data() {
+			stck[i] = item
+		}
+	}
+	// Copy a snapshot of the current storage to a new container
+	var (
+		storage Storage
+		pprof   [][]byte
```
Comment (0xmountaintop) [inline core/vm/logger.go]: do u mean `proof`?

File: core/types/l2trace.go [hunk 5]
```diff
@@ -10,10 +10,12 @@ import (
 
 // BlockResult contains block execution traces and results required for rollers.
 type BlockResult struct {
+	Block            *Block             `json:"block"`
```
Comment (0xmountaintop) [inline core/types/l2trace.go]: what's the pros of storing `Block` in `BlockResult`?

File: core/vm/logger.go [hunk 6]
```diff
@@ -177,7 +180,10 @@ func (l *StructLogger) CaptureState(pc uint64, op OpCode, gas, cost uint64, scop
 		}
 	}
 	// Copy a snapshot of the current storage to a new container
-	var storage Storage
+	var (
+		storage Storage
+		proof   [][]byte
```
Comment (0xmountaintop) [inline core/vm/logger.go]: prefer renaming to `storageProof`

File: core/vm/interpreter.go [hunk 7]
```diff
@@ -265,6 +265,10 @@ func (in *EVMInterpreter) Run(contract *Contract, input []byte, readOnly bool) (
 			in.returnData = res
 		}
 
+		if in.cfg.Debug {
+			in.cfg.Tracer.CaptureStateSpecial(pc, op, gasCopy, cost, callContext, in.returnData, in.evm.depth, err)
```
Comment (0xmountaintop) [inline core/vm/interpreter.go]: we can do 

`if in.cfg.Debug && op == SSTORE `

so that we can combine `CaptureStateSpecial` and `CaptureState`

File: core/vm/logger.go [hunk 8]
```diff
@@ -200,6 +206,72 @@ func (l *StructLogger) CaptureState(pc uint64, op OpCode, gas, cost uint64, scop
 			)
 			l.storage[contract.Address()][address] = value
 			storage = l.storage[contract.Address()].Copy()
+
+			proof, err = l.env.StateDB.GetStorageProof(contract.Address(), address)
+			if err != nil {
+				log.Warn("Failed to get proof", "contract address", contract.Address().String(), "key", address.String(), "err", err)
+			}
```
Comment (0xmountaintop) [inline core/vm/logger.go]: why only for `SSTORE` but no for `SLOAD`

File: core/vm/logger.go [hunk 9]
```diff
@@ -200,6 +206,72 @@ func (l *StructLogger) CaptureState(pc uint64, op OpCode, gas, cost uint64, scop
...
+		storage Storage
+		pprof   [][]byte
+	)
+	if !l.cfg.DisableStorage && stack.len() >= 2 {
+		// initialise new changed values storage container for this contract
+		// if not present.
+		if l.storage[contract.Address()] == nil {
+			l.storage[contract.Address()] = make(Storage)
+		}
+		// capture SLOAD opcodes and record the read entry in the local storage
```
Comment (0xmountaintop) [inline core/vm/logger.go]: according to your
```
	if op != SSTORE {
		return
	}
```

we cannot capture `SLOAD` here

Comment (mask-pp) [reply core/vm/logger.go]: SLOAD won't change storage, do we need add that?

File: core/vm/logger.go (hunk 8 above)
Comment (mask-pp) [reply core/vm/logger.go]: SLOAD won't change storage, do we need add that?

File: core/vm/interpreter.go (hunk 7 above)
Comment (mask-pp) [reply core/vm/interpreter.go]: The `op == SSTORE` is written in the function.

File: core/vm/logger.go (hunk 6 above)
Comment (mask-pp) [reply core/vm/logger.go:185]: Don't we need tx list and block property information?

File: core/types/l2trace.go (hunk 5 above)
Comment (mask-pp) [reply core/types/l2trace.go]: Don't we need tx list and block property information?

File: core/vm/logger.go (hunk 4 above)
Comment (mask-pp) [reply core/vm/logger.go]: Yes, the `storage path` is equivalent to `proof`.

File: core/blockchain.go (hunk 1 above)
Comment (mask-pp) [reply core/blockchain.go]: What's the difference?

File: core/vm/logger.go (hunk 6 above)
Comment (mask-pp) [reply core/vm/logger.go]: This is a temporary variable in a function. Are you sure need to change it?

Comment (mask-pp) [reply core/vm/logger.go:185]: This is a temporary variable in a function. Are you sure?

File: core/vm/logger.go (hunk 4 above)
Comment (0xmountaintop) [reply core/vm/logger.go]: yes but you have `pprof` instead of `proof` here.

IMO `pprof` has a special meaning for `performance profile` (if you search `pprof` on google)

File: core/blockchain.go (hunk 1 above)
Comment (0xmountaintop) [reply core/blockchain.go]: as I said, there's a side-effect, but your naming cannot implicate it.  

a function should not have any other side-effect if its name doesn't imply it

File: core/vm/logger.go (hunk 8 above)
Comment (0xmountaintop) [reply core/vm/logger.go]: > SLOAD won't change storage

it won't. but we still need proof for it (otherwise if we only have `sload` but no `sstore` in a block, how do you know the `storageProof`)

By my answer "doesn't mattter" in https://www.notion.so/L2geth-34ecc78de12c45348112c511f42f9e31?d=3be1aa921aaf47d8a34a04d22222855a#de9082e2cd3e4896b34ff8d40154ad7b I mean it doesn't matter before or after. but we need at least one

File: core/vm/logger.go (hunk 6 above)
Comment (0xmountaintop) [reply core/vm/logger.go]: I think so. because proof may also refer to "account proof". using `storageProof` can avoid ambiguity

File: core/vm/logger.go (hunk 8 above)
Comment (mask-pp) [reply core/vm/logger.go]: I see, `SLOAD` missed.

Comment (mask-pp) [reply core/vm/logger.go]: changed.

File: core/vm/logger.go (hunk 6 above)
Comment (mask-pp) [reply core/vm/logger.go]: Changed.

File: core/vm/logger.go (hunk 4 above)
Comment (mask-pp) [reply core/vm/logger.go]: changed.

**Comment by 0xmountaintop** [comment]
1. in core/types/l2trace.go#L94 and core/vm/logger.go#L73 it's still `Proof      *[]string          json:"storageProof,omitempty"` instead of `StorageProof      *[]string          json:"storageProof,omitempty"`

2. I understand that previously `op == SSTORE` is checked inside, but if we move it outside we can avoid duplicated codes (`CaptureState` vs `CaptureStateSpecial`) and no need for `log := StructLog{..., nil, ...}`

**Comment by 0xmountaintop** [comment]
I need to retake a look and define the doc&spec clearly.
I think we don't need path but storage_key and storage_value.

That is, I think https://www.notion.so/L2geth-34ecc78de12c45348112c511f42f9e31#2869187ad322433e8ddbc219cad38168 is not proper / well-defined.

Will keep u updated.

**Comment by 0xmountaintop** [comment]
I think we need to review https://www.notion.so/L2geth-34ecc78de12c45348112c511f42f9e31#2869187ad322433e8ddbc219cad38168 first before keep working on this PR

**Review by jules — APPROVED** [review]

**Comment by mask-pp** [comment]
> I think we need to review https://www.notion.so/L2geth-34ecc78de12c45348112c511f42f9e31#2869187ad322433e8ddbc219cad38168 first before keep working on this PR



> 1. in core/types/l2trace.go#L94 and core/vm/logger.go#L73 it's still `Proof      *[]string          json:"storageProof,omitempty"` instead of `StorageProof      *[]string          json:"storageProof,omitempty"`
> 2. I understand that previously `op == SSTORE` is checked inside, but if we move it outside we can avoid duplicated codes (`CaptureState` vs `CaptureStateSpecial`) and no need for `log := StructLog{..., nil, ...}`


> 

如果是键值对而不是路径，文档中

**Comment by 0xmountaintop** [comment]
Sorry just realize that we do need path.

and it turns out that your way (https://github.com/scroll-tech/go-ethereum/pull/20#discussion_r794269879) is better.

File: core/vm/logger_trace.go [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+	"github.com/scroll-tech/go-ethereum/common"
+	"github.com/scroll-tech/go-ethereum/common/hexutil"
+	"github.com/scroll-tech/go-ethereum/core/types"
+)
+
+type traceFunc func(l *StructLogger, scope *ScopeContext, extraData *types.ExtraData) error
+
+var (
+	// OpcodeExecs the map to load opcodes' trace funcs.
+	OpcodeExecs = map[OpCode][]traceFunc{
```
Comment (0xmountaintop) [inline core/vm/logger_trace.go:15]: nice

File: core/vm/logger_trace.go [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+func traceOriginProof(l *StructLogger, scope *ScopeContext, extraData *types.ExtraData) error {
+	address := l.env.Origin
+	proof, err := l.env.StateDB.GetProof(address)
+	if err == nil {
+		extraData.ProofList = append(extraData.ProofList, transferProof(proof))
+	}
+	return err
+}
+
+func transferProof(proof [][]byte) (res []string) {
```
Comment (0xmountaintop) [inline core/vm/logger_trace.go]: `encodeProof`?

File: core/vm/logger.go [hunk 12]
```diff
@@ -371,27 +411,30 @@ func FormatLogs(logs []StructLog) []types.StructLogRes {
...
-		if trace.Storage != nil {
+		if len(trace.Storage) != 0 {
 			storage := make(map[string]string)
 			for i, storageValue := range trace.Storage {
 				storage[fmt.Sprintf("%x", i)] = fmt.Sprintf("%x", storageValue)
 			}
 			formatted[index].Storage = &storage
 		}
+		if trace.ExtraData != nil {
+			formatted[index].ExtraData = trace.ExtraData.CleanExtraData()
```
Comment (0xmountaintop) [inline core/vm/logger.go]: I personally would prefer `CleanExtraData` -> `Seal` or `Finalize`, because `CleanExtraData` sounds like "remove ExtraData" 

But not a big problem. doesn't really matter

File: signer/fourbyte/4byte.go [hunk 13]
```diff
@@ -127472,7 +127472,7 @@ var __4byteJson = []byte(`{
 "de7bea74": "getThreshold(uint8)",
 "de7c2ed2": "TEAM_VESTING_DURATION()",
 "de7c99c3": "releasePrivateBonusTokens(address[],uint256[])",
-"de7cb327": "Proof()",
+"de7cb327": "StorageProof()",
```
Comment (0xmountaintop) [inline signer/fourbyte/4byte.go]: how to know whether we should change it? which part it's related to?

File: core/vm/logger.go [hunk 14]
```diff
@@ -200,15 +206,45 @@ func (l *StructLogger) CaptureState(pc uint64, op OpCode, gas, cost uint64, scop
...
+	execFuncList, ok := OpcodeExecs[op]
+	if !ok {
+		return
+	}
+	extraData := types.NewExtraData()
+	// execute trace func list.
+	for _, exec := range execFuncList {
+		if err = exec(l, scope, extraData); err != nil {
+			log.Error("Failed to trace data", "opcode", op.String(), "err", err)
+			break
```
Comment (0xmountaintop) [inline core/vm/logger.go]: maybe `continue` or just `return`? 

+ continue: try all the execFunc, omit failure
+ return: discard all if fail

---

`break` looks kind of "in the middle" of these two behaviors

File: core/vm/logger.go [hunk 15]
```diff
@@ -200,15 +206,45 @@ func (l *StructLogger) CaptureState(pc uint64, op OpCode, gas, cost uint64, scop
 			)
 			l.storage[contract.Address()][address] = value
 			storage = l.storage[contract.Address()].Copy()
+
+			extraData = types.NewExtraData()
+			if err := traceStorageProof(l, scope, extraData); err != nil {
+				log.Warn("Failed to get proof", "contract address", contract.Address().String(), "key", address.String(), "err", err)
+			}
```
Comment (0xmountaintop) [inline core/vm/logger.go:213]: what about we 
1 )   always traceStorageProof for `SLOAD` & `SSTORE`, and only traceStorageProofAfter for `SSTORE`

for now we can doing
2)   always only traceStorageProof for `SSTORE`, and traceStorageProofAfter for `SLOAD` & `SSTORE`

To me 1) looks more clean. but not a big problem, doesn't really matter

File: core/types/l2trace.go [hunk 16]
```diff
@@ -1,165 +1,78 @@
...
-	Op      string
-	Gas     uint64
-	GasCost uint64
-	Depth   uint
-	Error   string
-	Stack   []string
-	Memory  []string
-	Storage []string
+type ExtraData struct {
+	// CREATE: sender address
```
Comment (0xmountaintop) [inline core/types/l2trace.go]:
```suggestion
	// CREATE | CREATE2: sender address
```

File: core/blockchain.go [hunk 17]
```diff
@@ -1333,6 +1343,32 @@ func (bc *BlockChain) writeBlockWithState(block *types.Block, receipts []*types.
...
+			for i := range proof {
+				evmTrace.Proof[i] = hexutil.Encode(proof[i])
+			}
+		}
+		// Contract is called
+		if len(tx.Data()) != 0 && tx.To() != nil {
+			evmTrace.ByteCode = hexutil.Encode(state.GetCode(*tx.To()))
+		} else if tx.To() == nil { // Contract is created.
+			evmTrace.ByteCode = hexutil.Encode(tx.Data())
+		}
```
Comment (0xmountaintop) [inline core/blockchain.go:1371]: do we forget to deal with "For Transfer: tx.to address’s code_hash" ?

File: core/types/l2trace.go [hunk 18]
```diff
@@ -1,165 +1,78 @@
...
+	// CREATE: sender nonce
+	Nonce *uint64 `json:"nonce,omitempty"`
+	// CALL | CALLCODE | DELEGATECALL | STATICCALL: [tx.to address’s code_hash, stack.nth_last(1) address’s code_hash]
+	CodeHashList []common.Hash `json:"codeHashList,omitempty"`
+	// SSTORE | SLOAD: [storageProof]
+	// SELFDESTRUCT: [contract address’s accountProof, stack.nth_last(0) address’s accountProof]
+	// SELFBALANCE: [contract address’s accountProof]
+	// BALANCE | EXTCODEHASH: [stack.nth_last(0) address’s accountProof]
+	// CREATE | CREATE2: [created contract address’s accountProof]
+	// CALL | CALLCODE: [caller contract address’s accountProof, callee contract address’s accountProof, stack.nth_last(1) address’s accountProof]
```
Comment (0xmountaintop) [inline core/types/l2trace.go]: `callee contract address` is `stack.nth_last(1) address`, `stack.nth_last(1) address` is `callee contract address`. They are a same thing

File: core/vm/logger_trace.go [hunk 19]
```diff
@@ -0,0 +1,126 @@
...
+	"github.com/scroll-tech/go-ethereum/core/types"
+)
+
+type traceFunc func(l *StructLogger, scope *ScopeContext, extraData *types.ExtraData) error
+
+var (
+	// OpcodeExecs the map to load opcodes' trace funcs.
+	OpcodeExecs = map[OpCode][]traceFunc{
+		CALL:         {traceToAddressCodeHash, traceLastNAddressCodeHash(1), traceOriginProof, traceLastNAddressProof(1)},
+		CALLCODE:     {traceToAddressCodeHash, traceLastNAddressCodeHash(1), traceLastNAddressProof(0)},
```
Comment (0xmountaintop) [inline core/vm/logger_trace.go]:
```suggestion
		CALLCODE:     {traceToAddressCodeHash, traceLastNAddressCodeHash(1), traceLastNAddressProof(0)}, traceOriginProof, traceLastNAddressProof(1)},
```

File: signer/fourbyte/4byte.go (hunk 13 above)
Comment (mask-pp) [reply signer/fourbyte/4byte.go]: That's a mistake, will be changed.

File: core/types/l2trace_block.go [hunk 20]
```diff
@@ -0,0 +1,76 @@
...
+
+import (
+	"math/big"
+
+	"github.com/scroll-tech/go-ethereum/common"
+	"github.com/scroll-tech/go-ethereum/common/hexutil"
+	"github.com/scroll-tech/go-ethereum/params"
+)
+
+type TraceBlock struct {
```
Comment (icemelon) [inline core/types/l2trace_block.go]:
```suggestion
type BlockTrace struct {
```

File: core/types/l2trace_block.go [hunk 21]
```diff
@@ -0,0 +1,76 @@
...
+	Hash        common.Hash         `json:"hash"`
+	GasLimit    uint64              `json:"gasLimit"`
+	Difficulty  *big.Int            `json:"difficulty"`
+	BaseFee     *big.Int            `json:"baseFee"`
+	Coinbase    common.Address      `json:"coinbase"`
+	Time        uint64              `json:"time"`
+	Transaction []*TraceTransaction `json:"transaction"`
+}
+
+type TraceTransaction struct {
```
Comment (icemelon) [inline core/types/l2trace_block.go]:
```suggestion
type TransactionTrace struct {
```

Comment (icemelon) [reply core/types/l2trace_block.go]: Add a `TxType` field as we may need to use it in the future.

File: eth/api.go [hunk 22]
```diff
@@ -618,7 +618,10 @@ func NewPublicTraceAPI(eth *Ethereum) *PublicTraceAPI {
...
 
-// BlockResultByHash returns the blockResult by blockHash.
-func (api *PublicTraceAPI) BlockResultByHash(blockHash common.Hash) (*types.BlockResult, error) {
-	return rawdb.ReadBlockResult(api.e.chainDb, blockHash), nil
+// GetBlockResultByHash returns the blockResult by blockHash.
+func (api *PublicTraceAPI) GetBlockResultByHash(blockHash common.Hash) (*types.BlockResult, error) {
+	if blockResult := api.e.blockchain.GetBlockResultByHash(blockHash); blockResult != nil {
+		return blockResult, nil
+	}
+	return nil, fmt.Errorf("no block result found")
```
Comment (icemelon) [inline eth/api.go]:
```suggestion
	return nil, fmt.Errorf("No block result found")
```

File: eth/protocols/snap/handler.go [hunk 23]
```diff
@@ -341,7 +341,7 @@ func handleMessage(backend Backend, peer *Peer) error {
 				for _, blob := range proof.NodeList() {
 					proofs = append(proofs, blob)
 				}
-				// Proof terminates the reply as proofs are only added if a node
+				// StorageProof terminates the reply as proofs are only added if a node
```
Comment (icemelon) [inline eth/protocols/snap/handler.go]: revert the change?

File: core/vm/logger.go (hunk 14 above)
Comment (mask-pp) [reply core/vm/logger.go]: Good idea! continue to run maybe much better.

File: core/types/l2trace_block.go (hunk 21 above)
Comment (mask-pp) [reply core/types/l2trace_block.go]: OK

File: eth/protocols/snap/handler.go (hunk 23 above)
Comment (mask-pp) [reply eth/protocols/snap/handler.go]: yeah, should be reverted !

File: core/blockchain.go (hunk 17 above)
Comment (mask-pp) [reply core/blockchain.go:1371]: Yeah, it a little awkward, i need to think how to add gracefully.

File: core/types/l2trace.go (hunk 18 above)
Comment (mask-pp) [reply core/types/l2trace.go]: Yeah, the comment is a mistake, will be changed.

File: core/vm/logger_trace.go (hunk 11 above)
Comment (mask-pp) [reply core/vm/logger_trace.go]: Changed!

File: core/blockchain.go (hunk 17 above)
Comment (mask-pp) [reply core/blockchain.go:1371]: Tx has the tx.to address's code_hash when is a contract call.

File: core/types/l2trace.go [hunk 24]
```diff
@@ -1,165 +1,78 @@
...
-	"sort"
-	"strings"
-
-	"github.com/scroll-tech/go-ethereum/rlp"
+	"github.com/scroll-tech/go-ethereum/common"
 )
 
 // BlockResult contains block execution traces and results required for rollers.
 type BlockResult struct {
+	TraceBlock       *BlockTrace        `json:"block"`
```
Comment (icemelon) [inline core/types/l2trace.go]:
```suggestion
	BlockTrace       *BlockTrace        `json:"blockTrace"`
```

File: core/blockchain.go (hunk 17 above)
Comment (0xmountaintop) [reply core/blockchain.go:1371]: We actually also need `code_hash` for transfer (instead of "contract call")

but seems it's fixed and we can calculate it in  rust : https://eips.ethereum.org/EIPS/eip-1186

File: core/vm/logger_trace.go [hunk 25]
```diff
@@ -0,0 +1,126 @@
...
+}
+
+// traceNonce gets sender nonce
+func traceNonce(l *StructLogger, scope *ScopeContext, extraData *types.ExtraData) error {
+	nonce := l.env.StateDB.GetNonce(l.env.Origin)
+	extraData.Nonce = &nonce
+	return nil
+}
+
+// traceStorageProof get contract address's proof
```
Comment (0xmountaintop) [inline core/vm/logger_trace.go]:
```suggestion
// traceStorageProof get contract's storage proof at storage_address
```

File: core/vm/logger_trace.go [hunk 26]
```diff
@@ -0,0 +1,126 @@
...
+	contract := scope.Contract
+	// Get storage proof.
+	storageProof, err := l.env.StateDB.GetStorageProof(contract.Address(), address)
+	if err == nil {
+		extraData.ProofList = append(extraData.ProofList, encodeProof(storageProof))
+	}
+	return err
+}
+
+// traceContractProof gets the account address's proof
```
Comment (0xmountaintop) [inline core/vm/logger_trace.go]:
```suggestion
// traceContractProof gets the contract's account proof
```

File: core/vm/logger_trace.go [hunk 27]
```diff
@@ -0,0 +1,126 @@
...
+
+var (
+	// OpcodeExecs the map to load opcodes' trace funcs.
+	OpcodeExecs = map[OpCode][]traceFunc{
+		CALL:         {traceToAddressCodeHash, traceLastNAddressCodeHash(1), traceOriginProof, traceLastNAddressProof(1)},
+		CALLCODE:     {traceToAddressCodeHash, traceLastNAddressCodeHash(1), traceOriginProof, traceLastNAddressProof(1)},
+		DELEGATECALL: {traceToAddressCodeHash, traceLastNAddressCodeHash(1)},
+		STATICCALL:   {traceToAddressCodeHash, traceLastNAddressCodeHash(1)},
+		CREATE:       {traceSenderAddress, traceNonce},
+		CREATE2:      {traceSenderAddress},
```
Comment (0xmountaintop) [inline core/vm/logger_trace.go]: missing  "created contract address’s accountProof"?

File: core/vm/logger_trace.go [hunk 28]
```diff
@@ -0,0 +1,126 @@
...
+	"github.com/scroll-tech/go-ethereum/core/types"
+)
+
+type traceFunc func(l *StructLogger, scope *ScopeContext, extraData *types.ExtraData) error
+
+var (
+	// OpcodeExecs the map to load opcodes' trace funcs.
+	OpcodeExecs = map[OpCode][]traceFunc{
+		CALL:         {traceToAddressCodeHash, traceLastNAddressCodeHash(1), traceOriginProof, traceLastNAddressProof(1)},
+		CALLCODE:     {traceToAddressCodeHash, traceLastNAddressCodeHash(1), traceOriginProof, traceLastNAddressProof(1)},
```
Comment (0xmountaintop) [inline core/vm/logger_trace.go]: using `traceOriginProof` here can be wrong. caller may not be the tx sender.

we can retrieve caller_address using "ScopeContext.Contract.CallerAddress"

**Comment by 0xmountaintop** [comment]
If @ChuhanJin thinks it's urgent we can merge it now, but I think there are some mistakes in thi PR:
1. we want "caller's proof" instead of "origin (tx_sender)'s proof" in `CALL` & `CALLCODE`: https://github.com/scroll-tech/go-ethereum/pull/20#discussion_r809773939
2. we miss "created contract address’s accountProof" in `CREATE` and `CREATE2`: https://github.com/scroll-tech/go-ethereum/pull/20#discussion_r809773034

File: core/vm/logger_trace.go (hunk 27 above)
Comment (HAOYUatHZ) [inline core/vm/logger_trace.go:19]: And also sorry that, I made mistake in our [doc](https://www.notion.so/EVM-Trace-2e476641b6484b71a029c868516ad7b2): 
here for `CREATE` and `CREATE2` we need "callerAddress" "callerNnoce" instead of "senderAddress" "senderNonce".

we can use "ScopeContext.Contract.CallerAddress" to retrieve caller_address

Comment (HAOYUatHZ) [reply core/vm/logger_trace.go:19]: I also just update our doc

**Comment by ChuhanJin** [comment]
Can one of the admins verify this patch?

File: core/vm/logger_trace.go (hunk 27 above)
Comment (mask-pp) [reply core/vm/logger_trace.go]: Do both CREATE and CREATE2 need to be added?

File: core/vm/logger_trace.go (hunk 28 above)
Comment (mask-pp) [reply core/vm/logger_trace.go]: [This line](https://github.com/scroll-tech/go-ethereum/blob/8b16b4cefae82f7d3f73a37df9e04d7d356a7a23/core/evm.go#L70) can show origin is tx sender address.

But this's a mistake we only need contract caller proof, Don't tx's sender proof, that right?

File: core/vm/logger_trace.go (hunk 27 above)
Comment (0xmountaintop) [reply core/vm/logger_trace.go]: according to our https://www.notion.so/EVM-Trace-2e476641b6484b71a029c868516ad7b2, both `CREATE` and `CREATE2`

File: core/vm/logger_trace.go (hunk 28 above)
Comment (0xmountaintop) [reply core/vm/logger_trace.go]: Yes, that's what I mean. We need caller instead of sender.  sometime the caller is the sender, but sometimes not

**Review by 0xmountaintop — APPROVED** [review]
